src/sentry/integrations/gitlab/search.py

```diff
@@ -0,0 +1,51 @@
+from __future__ import absolute_import
+
+from rest_framework.response import Response
+
+from sentry.api.bases.organization import OrganizationEndpoint, OrganizationPermission
+from sentry.models import Integration
+
+
+class GitlabIssueSearchEndpoint(OrganizationEndpoint):
```

Comment:
@mnoble I have no idea how to write a good test for the search endpoint. Do you have any pointers?

Reply:
I'd say you'd want to cover a few different types of tests:

1. Happy paths
2. Request validation logic
3. Missing resources
4. Distributed System Issues (since this interacts with an external service)

### Happy Paths
These are basically, when everything about the request is correct, we expect to see these things.

1. `test_finds_external_issue_results`: mock the external request (or stubs the method) to return data the way GitLab's docs dictate. Make a request with `field=externalIssue`. Then ensure the response is a 200 and the data is in the shape you expect.
2. `test_finds_project_results`: Same as above, but with `field = project`.
3. `test_finds_no_external_issues_results`: mock the external request  (or stubs the method) to return whatever it looks like for GitLab to return no results (empty array probably). Ensure response is 200 and the data is in the shape you expect.
4. `test_finds_no_project_results`: Same as above, but with `field = project`

### Request Validations
When the request is wrong in some way, how does the endpoint react?

1. `test_missing_field`: Make request without `field` and ensure you receive a 400. You can ensure the message is correct, but imo, the status code is the more important part. The message may change, the status should not.
2. `test_missing_query`: Same as above, but with `query = None`.
3. `test_invalid_field`: Pass `field = asddasdsad` and ensure the endpoint responds appropriately.

### Missing Resources
What happens when the data that we expect, does not exist? For these, I would only worry about the resources you're looking up explicitly in the endpoint. Things like Organization are handled by the endpoint superclass and not the direct responsibility of this endpoint.

1. `test_missing_integration`: Pass an `integration_id` that doesn't exist. Ensure 404.
2. `test_missing_installation`: Remove the `OrganizationIntegration` and see how it responds. Currently looks like it'd do whatever the API defaults to when `get_installation` errors out.

### Distributed System Issues
What happens when the external service we're consuming is unavailable?

1. `test_gitlab_unavailable`: Make the "gitlab service" return a 503 by mocking it out with `responses`. Ensure the endpoint responds accordingly.
2. `test_search_issues_request_fails`: Make the `search_issues` request bomb out somehow. Ensure the endpoint responds accordingly.
3. `test_projects_request_fails`: Make the `project` request bomb out somehow. Ensure the endpoint responds accordingly.

The idea is to try to cover as many cases as you can think of. Imagine you're using the page this powers and all the ways you'd end up with an error message somehow. You won't catch them all, but we can anticipate some of them.

Reply:
:O oh my gosh. thank you so much!